File: apps/web/src/lib/pairing.ts
/**
 * Pairing a new device.
 *
 * # Which way the QR points, and why it is not arbitrary
 *
 * The **new** device displays, the **old** one scans. A QR code can be photographed by
 * construction: putting a secret in it would amount to publishing it. This one carries only an
 * ephemeral public key and a deposit address, both worthless to whoever intercepts them.
 *
 * The old device then seals the packet under the shared X25519 secret and deposits it on the
 * server, which only ever sees a blob: it holds neither of the two private halves.
 *
 * # What is not protected
 *
 * The channel's security is **physical**: it rests on the user only scanning the screen they are
 * holding. An attacker who shows them their own QR code gets paired, and no cryptography can
 * prevent it. That is WhatsApp's model, and Signal's.
 */
import { Api } from "./api";
import { fromBase64, toBase64 } from "./keys";

/** What the QR code encodes. None of these fields is secret. */
export interface PairingCode {
  id: Uint8Array;
  publicKey: Uint8Array;
}

/**
 * Encodes the offer into a compact string.
 *
 * The QR code is not always practical — a screen that cannot be shown, no camera on a desktop
 * computer. The same string can then be copied by hand, with exactly the same properties: it
 * contains nothing secret.
 */
export function encodePairingCode(code: PairingCode): string {
  const joined = new Uint8Array(code.id.length + code.publicKey.length);
  joined.set(code.id, 0);
  joined.set(code.publicKey, code.id.length);
  return toBase64(joined).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

export function decodePairingCode(text: string): PairingCode {
  const normalized = text.trim().replace(/-/g, "+").replace(/_/g, "/");
  const bytes = fromBase64(normalized.padEnd(Math.ceil(normalized.length / 4) * 4, "="));

  // 16 bytes of id + 32 of X25519 key. An unexpected length signals a truncated entry, not an
  // attack — but failing here beats sealing into the void.
  if (bytes.length !== 48) {
    throw new Error("Invalid or incomplete pairing code.");
  }

  return { id: bytes.slice(0, 16), publicKey: bytes.slice(16) };
}

export function depositPairing(api: Api, id: Uint8Array, payload: Uint8Array): Promise<unknown> {
  return api.depositPairing(id, payload);
}

/**
 * Waits for the original device to deposit the packet.
 *
 * The server does not notify: we poll. The window is short — the packet contains enough to take
 * control of the account, and the server expires it after five minutes.
 */
export async function awaitPairing(
  id: Uint8Array,
  signal: { cancelled: boolean },
): Promise<Uint8Array | null> {
  const deadline = Date.now() + 5 * 60 * 1000;

  while (!signal.cancelled && Date.now() < deadline) {
    const payload = await Api.claimPairing(id);
    if (payload) return payload;
    await new Promise((resolve) => setTimeout(resolve, 1500));
  }
  return null;
}

File: crates/crypto-core/src/pairing.rs
//! Pairing a new device, over a visual channel.
//!
//! # The problem
//!
//! A brand-new device needs the account root key to attest itself, and later to attest other
//! devices. That secret can neither travel in the clear through the server nor be shown on
//! screen a second time — a phrase read again is a phrase that can be photographed.
//!
//! # What the QR code means, and why it is not arbitrary
//!
//! The **new** device displays, the **old** one scans. So the QR holds no secret: only an
//! ephemeral public key and enough to identify the device. A QR is photographable by
//! construction; putting a secret in it would amount to publishing it.
//!
//! The old device then seals the packet under the shared X25519 secret and drops it on the
//! server, which sees only an opaque blob: it holds neither private half.
//!
//! # What this does not protect
//!
//! The channel's security is **physical**: it rests on the user scanning only the screen they
//! are holding. An attacker who shows them their own QR is paired, and no cryptography can
//! prevent it. This is the same model as WhatsApp and Signal.

use aes_gcm::aead::{Aead, KeyInit, Payload};
use aes_gcm::{Aes256Gcm, Nonce};
use hkdf::Hkdf;
use rand_core::RngCore;
use sha2::Sha256;
use x25519_dalek::{EphemeralSecret, PublicKey, StaticSecret};
use zeroize::Zeroize;

use crate::error::{CryptoError, Result};

/// Length of the pairing identifier. Random, it serves as the drop address on the server.
pub const PAIRING_ID_LEN: usize = 16;

const NONCE_LEN: usize = 12;
const INFO_SEAL: &[u8] = b"wac-pairing-seal-v1";
const INFO_CONFIRM: &[u8] = b"wac-pairing-confirm-v1";

/// What the new device publishes in its QR code.
///
/// Nothing secret: the ephemeral key is public, and the pairing identifier is only a drop
/// address. A third party photographing this QR gains nothing — they cannot open the packet,
/// lacking the private half.
pub struct PairingOffer {
    secret: EphemeralSecret,
    public: PublicKey,
    id: [u8; PAIRING_ID_LEN],
}

impl PairingOffer {
    pub fn generate() -> Self {
        let secret = EphemeralSecret::random_from_rng(rand_core::OsRng);
        let public = PublicKey::from(&secret);

        let mut id = [0u8; PAIRING_ID_LEN];
        rand_core::OsRng.fill_bytes(&mut id);

        Self { secret, public, id }
    }

    pub fn id(&self) -> [u8; PAIRING_ID_LEN] {
        self.id
    }

    pub fn public_key(&self) -> [u8; 32] {
        self.public.to_bytes()
    }

    /// Opens the packet dropped by the originating device.
    ///
    /// Consumes the offer: the ephemeral secret is destroyed after a single use, which rules
    /// out replaying an old packet against the same key.
    pub fn open(self, sealed: &[u8]) -> Result<Opened> {
        if sealed.len() < 32 + NONCE_LEN {
            return Err(CryptoError::Malformed("truncated pairing packet"));
        }

        let mut peer = [0u8; 32];
        peer.copy_from_slice(&sealed[..32]);
        let peer = PublicKey::from(peer);

        let shared = self.secret.diffie_hellman(&peer);
        let (key, confirmation) = derive(shared.as_bytes(), &peer, &self.public, &self.id);

        let cipher = Aes256Gcm::new_from_slice(&key)
            .map_err(|_| CryptoError::Malformed("invalid pairing key"))?;
        let nonce = Nonce::from_slice(&sealed[32..32 + NONCE_LEN]);

        let plaintext = cipher
            .decrypt(nonce, Payload { msg: &sealed[32 + NONCE_LEN..], aad: &self.id })
            .map_err(|_| CryptoError::Malformed("unreadable or tampered pairing packet"))?;

        Ok(Opened { plaintext, confirmation })
    }
}

pub struct Opened {
    pub plaintext: Vec<u8>,
    /// Short code to compare by eye on both screens.
    pub confirmation: String,
}

/// Seals a packet for the new device.
///
/// `offer_public` and `offer_id` come from the QR code, hence from an out-of-band channel the
/// server does not see. That is what makes the drop safe despite a hostile server.
pub fn seal(
    offer_public: &[u8],
    offer_id: &[u8],
    plaintext: &[u8],
) -> Result<(Vec<u8>, String)> {
    let offer_public: [u8; 32] = offer_public
        .try_into()
        .map_err(|_| CryptoError::Malformed("pairing key of invalid size"))?;
    let offer_public = PublicKey::from(offer_public);

    // `StaticSecret` rather than `EphemeralSecret`: we need the matching public key in the
    // message, and to derive twice from the same secret. It is wiped right after.
    let mut ours = StaticSecret::random_from_rng(rand_core::OsRng);
    let ours_public = PublicKey::from(&ours);
    let shared = ours.diffie_hellman(&offer_public);
    ours.zeroize();

    let (key, confirmation) = derive(shared.as_bytes(), &ours_public, &offer_public, offer_id);

    let cipher = Aes256Gcm::new_from_slice(&key)
        .map_err(|_| CryptoError::Malformed("invalid pairing key"))?;

    let mut nonce_bytes = [0u8; NONCE_LEN];
    rand_core::OsRng.fill_bytes(&mut nonce_bytes);
    let nonce = Nonce::from_slice(&nonce_bytes);

    let ciphertext = cipher
        .encrypt(nonce, Payload { msg: plaintext, aad: offer_id })
        .map_err(|_| CryptoError::Malformed("pairing encryption failed"))?;

    let mut out = Vec::with_capacity(32 + NONCE_LEN + ciphertext.len());
    out.extend_from_slice(ours_public.as_bytes());
    out.extend_from_slice(&nonce_bytes);
    out.extend_from_slice(&ciphertext);

    Ok((out, confirmation))
}

/// Derives the sealing key and the confirmation code.
///
/// Both public keys enter the derivation **in a fixed order** — the originating device's, then
/// the new device's. Without that order the two sides would compute different values; without
/// including them at all, an attacker could replay a shared secret obtained in another
/// exchange.
fn derive(
    shared: &[u8],
    sender: &PublicKey,
    receiver: &PublicKey,
    pairing_id: &[u8],
) -> ([u8; 32], String) {
    let mut transcript = Vec::with_capacity(80);
    transcript.extend_from_slice(sender.as_bytes());
    transcript.extend_from_slice(receiver.as_bytes());
    transcript.extend_from_slice(pairing_id);

    let hkdf = Hkdf::<Sha256>::new(Some(&transcript), shared);

    let mut key = [0u8; 32];
    hkdf.expand(INFO_SEAL, &mut key).expect("32 bytes valid for HKDF-SHA256");

    let mut digest = [0u8; 4];
    hkdf.expand(INFO_CONFIRM, &mut digest).expect("4 bytes valid for HKDF-SHA256");

    // Six digits: long enough that an accidental collision is unlikely, short enough that a
    // human actually compares it instead of clicking "yes" without looking.
    let code = u32::from_be_bytes(digest) % 1_000_000;
    (key, format!("{code:06}"))
}
